torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -232,7 +297,25 @@ void Reducer::mark_variable_ready_dense(VariableIndex index) {
       TORCH_INTERNAL_ASSERT(!grad.is_alias_of(bucket_view));
       TORCH_INTERNAL_ASSERT(grad.device() == bucket_view.device());
       TORCH_INTERNAL_ASSERT(grad.numel() == bucket_view.numel());
-      bucket_view.copy_(grad.view({-1}), /* non_blocking */ true);
+      // AccumulateGrad doesn't HAVE to obey the grad layout contract.
```

Comment:
@mrshenli are the diffs here ok?  grad is now divided while being copied into bucket_view, but it's still not being modified, so I think it's ok.
A more subtle thing is that with this PR's other diffs, grad may no longer be rowmajor contiguous.  I'm not sure how that affects `dist_autograd` in general if the callback ends up routing into the dist_autograd context.